File: gemv/performance/OpenACC.c
#include <stdio.h>
#include <stdlib.h>
#include <openacc.h>
#include <time.h> // Include this header


void gemv_parallel(int n, float* A, float* x, float* y) {
    int i, j;
    float sum;

    #pragma acc parallel loop private(j, sum) present(A[0:n*n], x[0:n], y[0:n])
    for (i = 0; i < n; i++) {
        sum = 0.0;
        #pragma acc loop reduction(+:sum)
        for (j = 0; j < n; j++) {
            sum += A[i * n + j] * x[j];
        }
        y[i] = sum;
    }
}

int main() {
    int n = 43346; // Size of the matrix and vectors
    float *A, *x, *y;
    clock_t start, end;
    double totalTime = 0, avgTime;

    // Allocate memory
    A = (float*)malloc(n * n * sizeof(float));
    x = (float*)malloc(n * sizeof(float));
    y = (float*)malloc(n * sizeof(float));

    // Initialize A and x with some values
    for (int i = 0; i < n; i++) {
        x[i] = 1.0; // Example initialization
        for (int j = 0; j < n; j++) {
            A[i * n + j] = (i + j) % 100; // Example initialization
        }
    }

    // Warmup run
    gemv_parallel(n, A, x, y);

    // Timing runs
    for (int i = 0; i < 10; i++) {
        start = clock();
        gemv_parallel(n, A, x, y);
        end = clock();
        totalTime += (double)(end - start) / CLOCKS_PER_SEC;
    }

    avgTime = totalTime / 10;
    printf("Average time taken for 10 runs: %f seconds\n", avgTime);

    // Cleanup
    free(A);
    free(x);
    free(y);

    return 0;
}

File: gemv/performance/OpenACC.f90
! program test_gemv_parallel
!     USE iso_c_binding, ONLY: C_LONG_LONG
!     implicit none
!     integer, parameter :: n = 31622
!     REAL(KIND=8) :: A(n, n), x(n), y(n)
!     INTEGER :: i
!     INTEGER :: count_max
!     INTEGER(C_LONG_LONG) :: start_count, end_count, count_diff, count_rate
!     REAL :: wall_time

!     ! Initialize matrix A and vector x with some values
!     A = 1.0
!     x = 1.0
!     y = 0.0

!     ! Warm-up call
!     call gemv_parallel(n, A, x, y)

!     ! Reset y for the actual runs
!     y = 0.0

!     ! Find average per call time
!     count_max = 10
!     count_diff = 0
!     i = 0
!     DO i = 1, count_max
!         CALL system_clock(count_rate=count_rate)
!         CALL system_clock(start_count)
!         CALL gemv_parallel(n, A, x, y)
!         CALL system_clock(end_count)
!         count_diff = count_diff + (end_count - start_count)
!     END DO


!     wall_time = REAL(count_diff) / REAL(count_rate) / REAL(count_max)
!     PRINT *, "Average time of ", count_max, " calls: ", wall_time, " seconds"

! end program test_gemv_parallel

! subroutine gemv_parallel(n, A, x, y)
!     implicit none
!     integer, intent(in) :: n
!     real, intent(in) :: A(n, n)
!     real, intent(in) :: x(n)
!     real, intent(out) :: y(n)
!     integer :: i, j
!     real :: sum

!     !$ACC PARALLEL LOOP
!     do i = 1, n
!         sum = 0.0
!         do j = 1, n
!             sum = sum + A(i, j) * x(j)
!         end do
!         y(i) = sum
!     end do
!     !$ACC END PARALLEL LOOP

! end subroutine gemv_parallel

program test_gemv_parallel
    USE iso_c_binding, ONLY: C_LONG_LONG
    implicit none
    integer, parameter :: n = 43346
    REAL(KIND=8) :: A(n, n), x(n), y(n)
    INTEGER :: i
    INTEGER :: count_max
    INTEGER(C_LONG_LONG) :: start_count, end_count, count_diff, count_rate
    REAL :: wall_time

    ! Initialize matrix A and vector x with some values
    A = 1.0
    x = 1.0
    y = 0.0

    ! Warm-up call
    call gemv_parallel(n, A, x, y)

    ! Reset y for the actual runs
    y = 0.0

    ! Find average per call time
    count_max = 10
    count_diff = 0

    ! Transfer data to device once
    !$acc data copyin(A, x) create(y)
    DO i = 1, count_max
        CALL system_clock(count_rate=count_rate)
        CALL system_clock(start_count)
        CALL gemv_parallel(n, A, x, y)
        CALL system_clock(end_count)
        count_diff = count_diff + (end_count - start_count)
    END DO
    !$acc end data

    wall_time = REAL(count_diff) / REAL(count_rate) / REAL(count_max)
    PRINT *, "Average time of ", count_max, " calls: ", wall_time, " seconds"

end program test_gemv_parallel

subroutine gemv_parallel(n, A, x, y)
    implicit none
    integer, intent(in) :: n
    real(KIND=8), intent(in) :: A(n, n)
    real(KIND=8), intent(in) :: x(n)
    real(KIND=8), intent(out) :: y(n)
    integer :: i, j
    real :: sum

    !$ACC PARALLEL LOOP
    do i = 1, n
        sum = 0.0
        do j = 1, n
            sum = sum + A(i, j) * x(j)
        end do
        y(i) = sum
    end do
    !$ACC END PARALLEL LOOP

end subroutine gemv_parallel
